examples/step-60/step-60.cc

```diff
@@ -0,0 +1,1077 @@
+    // search, and then reuse all of this for subsequent points.
+    //
+    // GridTools::compute_point_locations returns a tuple where the first
+    // element is a vector of cells containing the input points, in this
+    // case support_points. For refinement, this is the only information we need,
+    // and this is exactly what happens now.
+    //
+    // When we need to assemble a coupling matrix, however, we'll also need the
+    // reference location of each point to evaluate the basis functions of the
+    // embedding space. The other elements of the tuple returned by
+    // GridTools::compute_point_locations allow you to reconstruct, for each
+    // point, what cell contains it, and what is the location in the reference
+    // cell of the given point. Since this information is better grouped into
+    // cells, then this is what the algorithm returns: a tuple, containing a
+    // vector of all cells that have at least one point in them, together with a
+    // list of all reference points and their corresponding index in the
+    // original vector.
+    //
+    // In the following loop, we will be ignoring all returned objects except
+    // the first, identifying all cells contain at least one support point of
+    // the embedded space. This allows for a simple adaptive refinement strategy:
+    // refining these cells and their neighbors.
+    //
+    // Notice that we need to do some sanity checks, in the sense that we want
+    // to have an embedding grid which is well refined around the embedded grid,
+    // but where two consecutive support points lie either in the same cell, or
+    // in neighbor embedding cells.
+    //
+    // This is only possible if we ensure that the smallest cell size of the
+    // embedding grid is nonetheless bigger than the largest cell size of the
+    // embedded grid. Since users can modify both levels of refinements, as well
+    // as the amount of local refinement they want around the embedded grid, we
+    // make sure that the resulting meshes satisfy our requirements, and if this
+    // is not the case, we bail out with an exception.
+
+    for (unsigned int i=0; i<parameters.delta_refinement; ++i)
+      {
+        const auto point_locations = GridTools::compute_point_locations(*space_grid_tools_cache,
+                                     support_points);
+        const auto &cells = std::get<0>(point_locations);
+        for (auto cell : cells)
+          {
+            cell->set_refine_flag();
+            for (unsigned int face_no=0; face_no<GeometryInfo<spacedim>::faces_per_cell; ++face_no)
+              if (! cell->face(face_no)->at_boundary())
+                {
+                  auto neighbor = cell->neighbor(face_no);
+                  neighbor->set_refine_flag();
+                }
+          }
+        space_grid->execute_coarsening_and_refinement();
+        embedding_space_minimal_diameter = GridTools::minimal_cell_diameter(*space_grid);
```

Comment:
you can pull this line out of the loop and put it right before the following output statement